src/app/website/posts/zkp.markdown

```diff
@@ -0,0 +1,81 @@
+
+
+
+## What are Zero-knowledge Proofs?
+
+Now let’s consider a puzzle and see how you could convince anybody that you know a solution without revealing it. 
+
+
+### A Puzzle called 3-Coloring. 
+
+The 3-coloring puzzle can be described as follows. You are given a graph of nodes and edges (like in the figure below). The task is to find a "3-coloring" to the graph, which is a coloring of the nodes with three different colors in such a way that no two adjacent nodes have the same color. 
+
+![An uncolored graph.](/static/blog/zkp/uncolored-graph.png)
+
+![A colored graph.](/static/blog/zkp/colored-graph.png)
+
+### A Zero-knowledge Proof for the 3-coloring Puzzle. 
+
+We will construct a ZKP protocol for the 3-coloring puzzle. Before that, let’s quickly recall the two properties we are looking for in the protocol. 
+
+A ZKP protocol between you and someone -- call her  Verifier, must satisfy the following properties: 
+
+* If you are cheating (i.e., if you do not know a 3-coloring), then Verifier should be able to catch you -- this property is called  _soundness_
+
+* Verifier should not learn anything about the 3-coloring -- this property is called  _zero knowledgeness_
+
+Now, let's try to construct a ZKP protocol. Imagine the graph drawn on the floor in a closed space. Per the 3-coloring you know, you would place the corresponding colored balls on the nodes. Then you would completely cover the balls with inverted opaque bowls. 
+Then, Verifier comes and points at an edge of her choice. You would lift up the two bowls on either side of the edge. Verifier verifies that the balls revealed are of different colors. If they are not, you are caught cheating. 
+
+Notice that if you didnt really know a 3-coloring (meaning there is at least one edge with same colored balls on either side), you would get caught with some probability (namely, at least, $1/E$ where $E$ is the number of edges in the graph). Also, no further information about the 3-coloring is revealed to Verifier, since whatever Verifier viewed could be simulated by just picking two random, but differently-colored balls. Thus, this protocol provides some degree of soundness and perfect zero knowledgeness. 
```

Comment:
I think it might be helpful to separate this reasoning a little bit. Like 

Now that we've defined our protocol, we want to check that it satisfies *soundness* (the property that you can't cheat) and *zero-knowledgeness*.

## Soundness

If you didn't really know a 3-coloring, then for some two nodes connected by an edge, you put balls of the same color in step 1. That means in step 2, since Verifier picks an edge at random, the probability that they catch you is at least $1 / |E|$, where $|E|$ is the number of edges in the graph.

etc.

Reply:
Done.